File: src/main/java/dev/robothanzo/werewolf/service/GameActionService.java
package dev.robothanzo.werewolf.service;

import java.util.function.Consumer;

/**
 * Service for performing game-related actions such as resetting the game,
 * marking players as dead, reviving players, and updating game state.
 */
public interface GameActionService {
    /**
     * Resets the game session for a specific guild.
     *
     * @param guildId          the ID of the guild
     * @param statusCallback   callback for status messages
     * @param progressCallback callback for progress updates (0-100)
     * @throws Exception if an error occurs during the reset process
     */
    void resetGame(long guildId, Consumer<String> statusCallback, Consumer<Integer> progressCallback) throws Exception;

    /**
     * Marks a player as dead in the game session.
     *
     * @param guildId        the ID of the guild
     * @param userId         the ID of the user to mark as dead
     * @param allowLastWords whether the player is allowed to give last words
     */
    void markPlayerDead(long guildId, long userId, boolean allowLastWords);

    /**
     * Revives a player in the game session.
     *
     * @param guildId the ID of the guild
     * @param userId  the ID of the user to revive
     */
    void revivePlayer(long guildId, long userId);

    /**
     * Revives a player and assigns them a specific role.
     *
     * @param guildId the ID of the guild
     * @param userId  the ID of the user to revive
     * @param role    the role to assign to the player
     */
    void reviveRole(long guildId, long userId, String role);

    /**
     * Assigns the police (sheriff) status to a player.
     *
     * @param guildId the ID of the guild
     * @param userId  the ID of the user to make police
     */
    void setPolice(long guildId, long userId);

    /**
     * Broadcasts a progress update for a long-running action.
     *
     * @param guildId the ID of the guild
     * @param message the progress message
     * @param percent the progress percentage (0-100)
     */
    void broadcastProgress(long guildId, String message, Integer percent);
}

File: src/main/kotlin/dev/robothanzo/werewolf/service/GameActionService.kt
package dev.robothanzo.werewolf.service

import dev.robothanzo.werewolf.database.documents.Session

/**
 * Service for performing game-related actions such as resetting the game,
 * marking players as dead, reviving players, and updating game state.
 */
interface GameActionService {
    /**
     * Resets the game session.
     *
     * @param session          the session to reset
     * @param statusCallback   callback for status messages
     * @param progressCallback callback for progress updates (0-100)
     * @throws Exception if an error occurs during the reset process
     */
    @Throws(Exception::class)
    fun resetGame(session: Session, statusCallback: (String) -> Unit, progressCallback: (Int) -> Unit)

    /**
     * Revives a player in the game session.
     *
     * @param session the session containing the player
     * @param playerId the ID of the player to revive
     */
    fun revivePlayer(session: Session, playerId: Int)

    /**
     * Revives a player and assigns them a specific role.
     *
     * @param session the session containing the player
     * @param playerId the ID of the player to revive
     * @param role    the role to assign to the player
     */
    fun reviveRole(session: Session, playerId: Int, role: String)

    /**
     * Assigns the police (sheriff) status to a player.
     *
     * @param session the session containing the player
     * @param playerId the ID of the player to make police
     */
    fun setPolice(session: Session, playerId: Int)

    /**
     * Broadcasts a progress update for a long-running action.
     *
     * @param guildId the ID of the guild
     * @param message the progress message
     * @param percent the progress percentage (0-100)
     */
    fun broadcastProgress(guildId: Long, message: String?, percent: Int?)
}
